### PR TITLE
MCPWM Implementation update for ESP-IDF 5.0

### PR DESCRIPTION
**Pull request by SuperChamp234**
I am issuing this PR to update the MCPWM implementation for ESP-IDF 5.0.

As per pull request #88  , I have rewritten the MCPWM implementation for ESP-IDF 5.0. It make use of the `driver/mcpwm_prelude.h` now. 

This new MCPWM implementation requires the use of a `motor_handle_t` when trying to access the mcpwm object and perform actions.

Please review and let me know the changes, thank you.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -10,5 +12,5 @@ endif()
 
 idf_component_register(SRCS "${srcs}"
                        INCLUDE_DIRS "${include_dirs}"
-                       REQUIRES driver efuse esp32 esp_adc_cal esp_common esp_event 
-                                esp_rom freertos heap log soc spiffs)
\ No newline at end of file
+                       REQUIRES driver efuse esp_adc esp_common esp_event 
+                                esp_rom esp_timer freertos heap log soc spiffs)
```
Comment (RISHI27-dot) [inline CMakeLists.txt:16]: LGTM

File: include/motor_driver.h [hunk 2]
```diff
@@ -22,68 +22,36 @@
  * SOFTWARE.
  */
 
-#ifndef MOTOR_DRIVER_H
-#define MOTOR_DRIVER_H
+#pragma once
```
Comment (RISHI27-dot) [inline include/motor_driver.h:25]: Use header guards instead of pragma.

File: src/mcpwm_interface.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <sys/cdefs.h>
```
Comment (RISHI27-dot) [inline src/mcpwm_interface.c:27]: Remove these 3 included headers, as they are not required.

File: src/mcpwm_interface.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+#include "esp_log.h"
```
Comment (RISHI27-dot) [inline src/mcpwm_interface.c:29]: Including this is not required.

File: src/motor_driver.c [hunk 5]
```diff
@@ -22,151 +22,51 @@
  * SOFTWARE.
  */
 
+#include "esp_log.h"
+#include "esp_check.h"
```
Comment (RISHI27-dot) [inline src/motor_driver.c:26]: Remove these 2 included headers, as they are not required.

File: src/motor_driver.c [hunk 6]
```diff
@@ -22,151 +22,51 @@
...
-
-        return ESP_OK;
-    }
-    else
-    {
-        ESP_LOGE(TAG_MOTOR_DRIVER, "invalid motor direction selected");
-        return ESP_FAIL;
-    }
+#include "utils.h"
+#include "pin_defs.h"
```
Comment (RISHI27-dot) [inline src/motor_driver.c:29]: Including this is not required.

File: .gitignore [hunk 7]
```diff
@@ -75,9 +78,5 @@ dkms.conf
 # Executables
 
 ### vscode ###
-.vscode/*
-!.vscode/settings.json
-!.vscode/tasks.json
```
Comment (VedantParanjape) [inline .gitignore:80]: why are you removing all these?

File: CMakeLists.txt (hunk 1 above)
Comment (VedantParanjape) [reply CMakeLists.txt:16]: LGTM

File: examples/motor_driver_normal/main/main.c [hunk 8]
```diff
@@ -27,27 +27,19 @@
 #include "freertos/task.h"
 #include "sra_board.h"
 
-static const char* TAG_SWITCH_CONTROL_MOTOR = "switch_control";
```
Comment (VedantParanjape) [inline examples/motor_driver_normal/main/main.c:30]: Please use the tag char array to print logs, it's an unsaid rule in esp-idf.

File: examples/motor_driver_normal/main/main.c [hunk 9]
```diff
@@ -27,27 +27,19 @@
...
-        vTaskDelay(2000 / portTICK_PERIOD_MS);
-
-        // Rotate Motors A_0 and A_1 in BACKWARD direction
-        set_motor_speed(MOTOR_A_0, MOTOR_BACKWARD, 80);
-		ESP_LOGI(TAG_SWITCH_CONTROL_MOTOR, "MOTOR_A_0 BACKWARD");
-        set_motor_speed(MOTOR_A_1, MOTOR_BACKWARD, 80);
-		ESP_LOGI(TAG_SWITCH_CONTROL_MOTOR, "MOTOR_A_1 BACKWARD");
-
-        vTaskDelay(2000 / portTICK_PERIOD_MS);
+    motor_handle_t motor1;
```
Comment (VedantParanjape) [inline examples/motor_driver_normal/main/main.c:32]: indent nicely like before, add comments as well.

File: include/motor_driver.h (hunk 2 above)
Comment (VedantParanjape) [reply include/motor_driver.h:25]: +1

File: src/i2cdev.c
Comment (VedantParanjape) [inline src/i2cdev.c:1]: LGTM!

**Review by VedantParanjape — CHANGES_REQUESTED** [review]
I will want to review once more, after Rishi finished the review.

File: src/motor_driver.c (hunk 6 above)
Comment (SuperChamp234) [reply src/motor_driver.c:29]: This is required as we use the pindefs for creation of mcpwm config, and we use the map function from utils for mapping the PWM.

File: .gitignore (hunk 7 above)
Comment (SuperChamp234) [reply .gitignore]: Do we need the `.vscode` files in the repository?

File: include/motor_driver.h (hunk 2 above)
Comment (SuperChamp234) [reply include/motor_driver.h]: Changes made

File: src/mcpwm_interface.c (hunk 3 above)
Comment (SuperChamp234) [reply src/mcpwm_interface.c]: Changes made 👍🏽

File: src/mcpwm_interface.c (hunk 4 above)
Comment (SuperChamp234) [reply src/mcpwm_interface.c]: Changes made 👍🏽

File: src/motor_driver.c (hunk 5 above)
Comment (SuperChamp234) [reply src/motor_driver.c]: Changed made 👍🏽

File: examples/motor_driver_normal/main/main.c (hunk 8 above)
Comment (SuperChamp234) [reply examples/motor_driver_normal/main/main.c:30]: Change made 👍🏽

File: examples/motor_driver_normal/main/main.c (hunk 9 above)
Comment (SuperChamp234) [reply examples/motor_driver_normal/main/main.c:34]: Change made 👍🏽

File: include/sra_board.h
Comment (RISHI27-dot) [inline include/sra_board.h:1]: LGTM!

File: CMakeLists.txt
Comment (RISHI27-dot) [inline CMakeLists.txt:1]: LGTM

File: examples/motor_driver_normal/main/main.c [hunk 10]
```diff
@@ -27,27 +27,21 @@
 #include "freertos/task.h"
 #include "sra_board.h"
 
-static const char* TAG_SWITCH_CONTROL_MOTOR = "switch_control";
+static const char* TAG = "MOTOR";
```
Comment (RISHI27-dot) [inline examples/motor_driver_normal/main/main.c:30]: LGTM

File: include/motor_driver.h [hunk 11]
```diff
@@ -25,65 +25,46 @@
...
- * 
- * @param id for id=a, returns mode of motor driver a
- * @return int returns mode of motor driver A, 1 = Parallel mode, 2 = Normal mode
+ *  @brief Enable motor driver
+ *  @param motor_handle_t Motor object handle
+ *  @param motor_id set it as MOTOR_A_0, MOTOR_A_1 to select the appropriate motor to bind to the handle.
+ *  @return
+ *    - ESP_OK: Success
+ * - ESP_ERR_INVALID_ARG: Invalid argument
+ * - ESP_FAIL: Failed
```
Comment (RISHI27-dot) [inline include/motor_driver.h:47]: Indent this properly

File: src/motor_driver.c [hunk 12]
```diff
@@ -23,150 +23,55 @@
...
+		ESP_LOGE(TAG, "Invalid motor id");
+		return ESP_FAIL;
+	}
+	motor_mcpwm_config_t motor_mcpwm_config;
+	motor_mcpwm_config.group_id = 0;
+	motor_mcpwm_config.resolution_hz = MCPWM_RESOLUTION; // 10MHz
+	ESP_ERROR_CHECK(motor_new_mcpwm_device(&motor_config, &motor_mcpwm_config, motor));
+	ESP_ERROR_CHECK((*motor)->enable(*motor));
+	enabled_motor_driver_flag = 1;
+	return ESP_OK;
```
Comment (RISHI27-dot) [inline src/motor_driver.c:53]: LGTM.

File: src/motor_driver.c
Comment (RISHI27-dot) [inline src/motor_driver.c:1]: LGTM!

File: include/mcpwm_interface.h
Comment (RISHI27-dot) [inline include/mcpwm_interface.h:1]: LGTM.

File: src/mcpwm_interface.c
Comment (RISHI27-dot) [inline src/mcpwm_interface.c:1]: LGTM!

File: include/motor_driver.h [hunk 13]
```diff
@@ -25,65 +25,46 @@
...
 /**
- * @brief Reads mode of motor driver A
- * 
- * @return int returns mode of motor driver A, 1 = Parallel mode, 2 = Normal mode
+ * @brief Checks if the motor driver is engaged.
+ *
+ * @return true if the motor driver is engaged, false otherwise.
  */
-int read_motor_driver_mode_a();
+int get_motor_driver_status();
```
Comment (RISHI27-dot) [inline include/motor_driver.h:68]: This looks good.

File: include/motor_driver.h [hunk 14]
```diff
@@ -25,65 +25,46 @@
...
 
-#endif
+#endif
```
Comment (RISHI27-dot) [inline include/motor_driver.h:70]: Check for redundant addition of new lines.

File: src/bar_graph.c
Comment (RISHI27-dot) [inline src/bar_graph.c:1]: LGTM!

File: src/motor_driver.c (hunk 6 above)
Comment (RISHI27-dot) [reply src/motor_driver.c:27]: ok

File: include/utils.h [hunk 15]
```diff
@@ -25,6 +25,8 @@
 #ifndef _UTILS_H
 #define _UTILS_H
 
+#include "esp_log.h"
```
Comment (RISHI27-dot) [inline include/utils.h:28]: Remove this include if not required.

**Review by RISHI27-dot — CHANGES_REQUESTED** [review]

**Review by RISHI27-dot — APPROVED** [review]
@VedantParanjape The Wall-E bot successfully achieves self-balancing through the implemented MCPWM driver, and I have completed my review. Please add your reviews  so that we can proceed with merging.

File: src/motor_driver.c [hunk 16]
```diff
@@ -23,150 +23,55 @@
...
-
-        return ESP_OK;
-    }
-    else
-    {
-        ESP_LOGE(TAG_MOTOR_DRIVER, "invalid motor direction selected");
-        return ESP_FAIL;
-    }
+esp_err_t set_motor_speed(motor_handle_t motor, int direction, float speed) {
+	speed = (uint32_t)map(speed, 0, 100, 0, MCPWM_RESOLUTION / MCPWM_FREQ);
```
Comment (RISHI27-dot) [inline src/motor_driver.c:57]: @SuperChamp234  Please add a description of how this mapping is done in the commit message.

**Pull request merged**